packages/eslint-plugin/src/rules/no-non-null-asserted-nullish-coalescing.ts

```diff
@@ -0,0 +1,44 @@
+import { TSESTree, TSESLint } from '@typescript-eslint/experimental-utils';
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'no-non-null-asserted-nullish-coalescing',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallows using a non-null assertion in the left operand of the nullish coalescing operator',
+      category: 'Possible Errors',
+      recommended: false,
+      suggestion: true,
+    },
+    messages: {
+      noNonNullAssertedNullishCoalescing:
+        'The nullish coalescing operator is designed to handle undefined and null - using a non-null assertion is not needed.',
+      suggestRemovingNonNull: 'You should remove the non-null assertion.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      'LogicalExpression[operator = "??"] > TSNonNullExpression.left'(
+        node: TSESTree.TSNonNullExpression,
+      ): void {
+        context.report({
+          node,
+          messageId: 'noNonNullAssertedNullishCoalescing',
+          // use a suggestion instead of a fixer, because this can obviously break type checks
+          suggest: [
+            {
+              messageId: 'suggestRemovingNonNull',
+              fix(fixer): TSESLint.RuleFix {
+                return fixer.removeRange([node.range[1] - 1, node.range[1]]);
+              },
+            },
+          ],
```

Comment:
aside from the use-before-assigned case I mentioned - what other cases can you think of that could break code?

I think that thanks to the nullish coalesce, it should actually be 100% save to remove the assertion.

Reply:
Your question is about why I choose a suggestion instead of a fixer and what are those other cases that could break, isn't it?

In the following example removing the non-null-assertion will break code. If the non-null-assertion is removed typescript reports this error: `Type 'string | false' is not assignable to type 'string'`. 

```ts
function test(x?: string): string {
    const bar = x! ?? false; // type analysis reports `bar` has type `string`
    //          x  ?? false; // type analysis reports `bar` has type `string | false`
    return bar;
}
```

Reply:
ah - I understand now. I didn't realise that TypeScript did something silly here.. I mean it makes sense from a types POV, but it still seems weird from a control-flow perspective.

Reply:
it would probably be worth including a note in the docs about why we use a suggestion fixer - I don't think this would be obvious to people.

It would stave off some future issues if we documented why we made this choice.

Reply:
> it would probably be worth including a note in the docs about why we use a suggestion fixer - I don't think this would be obvious to people.
> 
> It would stave off some future issues if we documented why we made this choice.

Done